server/src-lib/Hasura/GraphQL/Resolve/Action.hs

```diff
@@ -210,7 +210,7 @@ resolveAsyncActionQuery userInfo selectOpCtx field = do
 
   annotatedFields <- fmap (map (first FieldName)) $ withSelSet (_fSelSet field) $ \fld ->
     case _fName fld of
-      "__typename" -> return $ RS.FExp $ G.unName $ G.unNamedType $ _fType fld
+      "__typename" -> return $ RS.FExp $ G.unName $ G.unNamedType $ _fType field
       "output"     -> do
```

Comment:
Using `fld` type here is a bug which always results in `String` for `__typename` field.